### PR TITLE
api-ErrorBudgetBurn fires before cluster api is Ready

### PR DESCRIPTION
**Pull request by anispate**
As part of [OSD-16950](https://issues.redhat.com/browse/OSD-16950), Currently, api-ErrorBudgetBurn fires before cluster api are Ready therefore, we are fixing that aspect as part of this PR.

It is making sure the status of the ClusterVersionAvailable="true" before firing the alert.

### CODE REVIEW

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 1]
```diff
@@ -67,6 +68,16 @@ func (s *ClusterUrlMonitorReconciler) EnsurePrometheusRuleExists(clusterUrlMonit
 	return utilreconcile.ContinueReconcile()
 }
 
+// FindStatusCondition finds the conditionType in conditions.
+func (s *ClusterUrlMonitorReconciler) FindStatusCondition(conditions []metav1.Condition, conditionType string) *metav1.Condition {
```
Comment (sam-nguyen7) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:72]: Are there any benefits to have our own implementation of FindStatusCondition as opposed to using https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#FindStatusCondition?

Comment (sam-nguyen7) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:72]: We can simplify this logic by having one helper function determine if cluster is available (eg. https://github.com/openshift/hypershift/pull/2147/files#diff-3d4ffda0834e09a59a4d510523932e20b9234bcb9816e40258a0fb32a87a7006R187) then requeue when !isClusterVersionAvailable

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 2]
```diff
@@ -90,7 +101,13 @@ func (s *ClusterUrlMonitorReconciler) EnsureServiceMonitorExists(clusterUrlMonit
 			return utilreconcile.RequeueReconcileWith(err)
 		}
 	}
-
+	condition, err := s.FindStatusCondition(hcp.Status.Conditions, string(metav1.ClusterVersionAvailable))
```
Comment (sam-nguyen7) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:104]: `FindStatusCondition` returns `*metav1.Condition`, but here you're expecting `condition, err`

**Review by sam-nguyen7 — CHANGES_REQUESTED** [review]
@anispate Please help add a description for this PR as well. Thanks!

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 3]
```diff
@@ -90,6 +101,16 @@ func (s *ClusterUrlMonitorReconciler) EnsureServiceMonitorExists(clusterUrlMonit
 			return utilreconcile.RequeueReconcileWith(err)
 		}
 	}
+	condition := s.FindStatusCondition(hcp.Status.Conditions, string(metav1.ClusterVersionAvailable))
+	if condition.Status = nil  {
+		return utilreconcile.RequeueReconcileWith(err)
+	}else{
+		condition != nil && condition.Status == metav1.ConditionTrue
+	}
```
Comment (fahlmant) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:109]: What does this block do?

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 4]
```diff
@@ -90,6 +101,16 @@ func (s *ClusterUrlMonitorReconciler) EnsureServiceMonitorExists(clusterUrlMonit
 			return utilreconcile.RequeueReconcileWith(err)
 		}
 	}
+	condition := s.FindStatusCondition(hcp.Status.Conditions, string(metav1.ClusterVersionAvailable))
+	if condition.Status = nil  {
+		return utilreconcile.RequeueReconcileWith(err)
```
Comment (fahlmant) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:106]: What's the expected err to return here?

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 5]
```diff
@@ -90,6 +101,16 @@ func (s *ClusterUrlMonitorReconciler) EnsureServiceMonitorExists(clusterUrlMonit
...
+	}else{
+		condition != nil && condition.Status == metav1.ConditionTrue
+	}
+
+	if condition.Status != metav1.ConditionTrue{
+		return utilreconcile.RequeueReconcileWith(err)
```
Comment (fahlmant) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:112]: What's the expected err to return here?

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 6]
```diff
@@ -67,6 +68,16 @@ func (s *ClusterUrlMonitorReconciler) EnsurePrometheusRuleExists(clusterUrlMonit
 	return utilreconcile.ContinueReconcile()
 }
 
+// FindStatusCondition finds the conditionType in conditions.
+func (s *ClusterUrlMonitorReconciler) FindStatusCondition(conditions []metav1.Condition, conditionType string) *metav1.Condition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
```
Comment (fahlmant) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:77]:
```suggestion
	for _,condition := range conditions {
		if condition.Type == conditionType {
			return &condition
		}
	}
```

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 7]
```diff
@@ -90,6 +101,16 @@ func (s *ClusterUrlMonitorReconciler) EnsureServiceMonitorExists(clusterUrlMonit
 			return utilreconcile.RequeueReconcileWith(err)
 		}
 	}
+	condition := s.FindStatusCondition(hcp.Status.Conditions, string(metav1.ClusterVersionAvailable))
+	if condition.Status = nil  {
```
Comment (fahlmant) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:105]:
```suggestion
	if condition.Status == nil  {
```

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 8]
```diff
@@ -84,6 +100,9 @@ func (s *ClusterUrlMonitorReconciler) EnsureServiceMonitorExists(clusterUrlMonit
 		if err != nil {
 			return utilreconcile.RequeueReconcileWith(err)
 		}
+		if !isClusterVersionAvailable(clusterUrl) {
+			return utilreconcile.RequeueReconcileWith(err)
```
Comment (fahlmant) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:104]: Wouldn't the `err` be nil here?

**Comment by fahlmant** [comment]
@anispate can you update the description to explain what the code change you're making does and how it fixes the issue?

**Review by mjlshen — COMMENTED** [review]
/lgtm

**Review by sam-nguyen7 — COMMENTED** [review]
/approve

**Pull request merged**